### PR TITLE
Add visible param to TabBar

### PR DESCRIPTION
**Pull request by microwavesafe**
This implements the ability to hide the tabBar by setting visible = false

It can be turned on in the navigationOptions
```
    static navigationOptions = {
        tabBar: ({ state }) => ({
            label: 'Messaging',
            visible: false,
        }),
    };
```
or accessed through SetParams (when #132 or similar functionality is implemented). See #136 for discussion.

### CODE REVIEW

File: src/views/TabView/TabView.js [hunk 1]
```diff
@@ -157,7 +157,10 @@ class TabView extends PureComponent<void, Props, void> {
     let renderHeader;
     let renderFooter;
 
-    if (typeof tabBarComponent !== 'undefined') {
+    const { state } = this.props.navigation;
+    const tabBar = this.props.router.getScreenConfig(this.props.childNavigationProps[state.routes[state.index].routeName], 'tabBar');
```
Comment (satya164) [inline src/views/TabView/TabView.js]: I think this should be `this.props.childNavigationProps[state.routes[state.index].key]`

reference - https://github.com/react-community/react-navigation/blob/master/src/withCachedChildNavigation.js#L51

Comment (microwavesafe) [inline src/views/TabView/TabView.js]: Ah yes, it was working on mine because they are the same value. I will update.

**Comment by microwavesafe** [comment]
Requested changes commited

**Comment by microwavesafe** [comment]
I guess we want to add something to the docs as well, but alongside my #132 pull request I can turn the tab bar on and off using this code in the navigationOptions
```
   static navigationOptions = {
        tabBar: ({ state }) => ({
            label: 'Messaging',
                visible: typeof state.params === 'undefined' ? true : state.params.visible,
        }),
    };
```
and these functions to turn them on and off with the keyboard
```
    _keyboardDidShow = () => {
        this.props.navigation.setParams({ visible: false })
    };

    _keyboardDidHide = () => {
        this.props.navigation.setParams({ visible: true })
    };
```

File: src/views/TabView/TabView.js [hunk 2]
```diff
@@ -157,7 +157,10 @@ class TabView extends PureComponent<void, Props, void> {
     let renderHeader;
     let renderFooter;
 
-    if (typeof tabBarComponent !== 'undefined') {
+    const { state } = this.props.navigation;
+    const tabBar = this.props.router.getScreenConfig(this.props.childNavigationProps[state.routes[state.index].key], 'tabBar');
+
+    if ((typeof tabBarComponent !== 'undefined') && (tabBar.visible !== false)) {
```
Comment (ericvicenti) [inline src/views/TabView/TabView.js]: You can just do `tabBarComponent !== undefined`

**Comment by jhilden** [comment]
Thanks for this PR @microwavesafe   This is exactly what we need.  Let me know if I can help in any way to get this merged.

**Comment by microwavesafe** [comment]
The documentation is a bit sparse but hopefully it gets the point across!

File: docs/guides/Screen-Nav-Options.md [hunk 3]
```diff
@@ -96,7 +96,23 @@ The 2nd argument passed to the function are the default values for the `header`
 
 ## Tab Navigation Options
 
-Coming Soon
+```js
+class TabScreen extends React.Component {
+  static navigationOptions = {
+    label: 'Tab Label',
```
Comment (jhilden) [inline docs/guides/Screen-Nav-Options.md]: I'm not 100% sure, but don't these properties (`lable`, `icon`, `visible`) have to be nested into another property `tabBar`?

Comment (satya164) [inline docs/guides/Screen-Nav-Options.md]: you're right. @microwavesafe can you change this?

**Comment by microwavesafe** [comment]
@jhilden good catch, new docs committed.

**Comment by microwavesafe** [comment]
I've just committed another change. I was playing with a stack navigator underneath a tab navigator. Doing this you get a situation where the tab navigator tries to get the tabBar options from a stack navigator page, leading to a null error. The _renderIcon function already deals with this correctly, but my function caused a null error.

I'm not sure what we want to do as a default behaviour here. My last commit will close the entire tab bar. Previously it would have rendered the tab bar, but the current icon would disappear. Which behaviour would you prefer?

**Comment by microwavesafe** [comment]
I should add to display the tab bar when the page is a stack navigator, you just have to add the tabBar options as you would with a normal tab component.

File: src/views/TabView/TabView.js [hunk 4]
```diff
@@ -157,7 +157,10 @@ class TabView extends PureComponent<void, Props, void> {
     let renderHeader;
     let renderFooter;
 
-    if (typeof tabBarComponent !== 'undefined') {
+    const { state } = this.props.navigation;
+    const tabBar = this.props.router.getScreenConfig(this.props.childNavigationProps[state.routes[state.index].key], 'tabBar');
+
+    if ((tabBarComponent !== 'undefined') && (tabBar) && (tabBar.visible !== false)) {
```
Comment (satya164) [inline src/views/TabView/TabView.js]: `tabBarComponent !== 'undefined'` isn't correct, do `tabBarComponent !== undefined`

nit: remove unnecessary parenthesis

**Comment by satya164** [comment]
Can you show the stack trace of the error you get?

File: src/views/TabView/TabView.js [hunk 5]
```diff
@@ -157,7 +157,10 @@ class TabView extends PureComponent<void, Props, void> {
     let renderHeader;
     let renderFooter;
 
-    if (typeof tabBarComponent !== 'undefined') {
+    const { state } = this.props.navigation;
+    const tabBar = this.props.router.getScreenConfig(this.props.childNavigationProps[state.routes[state.index].key], 'tabBar');
+
+    if ((tabBarComponent !== undefined) && (tabBar) && (tabBar.visible !== false)) {
```
Comment (satya164) [inline src/views/TabView/TabView.js]: Please remove the unnecessary parenthesis so we can merge

**Comment by microwavesafe** [comment]
I realise that strict equality and inequality has a higher operator precedence than logical AND so I can remove all of the internal parenthesis, but I would have thought this would make the code harder to read?

Anyway can you clarify if you want
```
if ((tabBarComponent !== undefined) && tabBar && (tabBar.visible !== false))
```
OR
```
if (tabBarComponent !== undefined && tabBar && tabBar.visible !== false)
```
so I don't have to commit again if I get it wrong.

**Comment by satya164** [comment]
Too many parenthesis are harder to read :D

But anyways, yeah that's correct

**Comment by microwavesafe** [comment]
Which one is correct, the first or second one?

**Comment by satya164** [comment]
`if (tabBarComponent !== undefined && tabBar && tabBar.visible !== false)`

**Comment by microwavesafe** [comment]
Thought so but wanted to be sure! I'll do it now.

**Comment by starchand** [comment]
Can we update docs at https://reactnavigation.org/docs/navigators/tab
I'm confused how to hide the tabbar.

**Comment by KabirKang** [comment]
@microwavesafe what about if you nest a stacknavigator in a tabnavigator and u want to make it so that only some of the screens in the underlying stacknav hide the tabs? this isn't working for me. i set up a function to get the tabnavigator from a context object and the use setparams on it, but that isn't working. whats another way to selectively hide tabs from within the lower stacknav?

**Comment by microwavesafe** [comment]
I have this
```
    static navigationOptions = {
        title: 'Messaging',
        tabBar: ({ state }) => ({
            label: 'Messaging',
            icon: ({ tintColor, focused }) =>
                <MessagingIconBadge
                    MainElement={
                        <Icons
                        name={'message-text'}
                        size={32}
                        style={{ color: tintColor }}
                        />
                    }
                />,
            visible: !state.params ? true : state.params.visible,
        }),
    };
```
and these to set it within the same component

```
this.props.navigation.setParams({ visible: false })
this.props.navigation.setParams({ visible: true })
```

as long as you are accessing the correct navigation, I assume this would work elsewhere.

**Comment by KabirKang** [comment]
@microwavesafe but the navigationOptions on the component would be affecting the stack navigator that's nested in the tab navigator, not the tab navigator itself right? these are the navigation options on the ProfileScreen I have that are within the stack nav:

`static navigationOptions = {
    header: ({state: {params = {}}}) => {
      const title = (typeof params.teacherName === 'undefined') ? '' : params.teacherName;
      return {
        title
      }
    }
  }`

Adding a tabBar property here doesn't seem to work. Trying to explore other ideas like maintaining nav state with redux but i dont know if i'm overthinking it.

**Comment by KabirKang** [comment]
The tabs:

`this.TabNav = TabNavigator(
      HOME_TABS_ROUTER,
      {
        initialRouteName: this.initialTab,
        tabBarPosition: 'bottom',
        tabBarComponent: TabView.TabBarBottom,
        order: [HOME_TABS.FIND_TEACHER_TAB, HOME_TABS.CHAT_TAB, HOME_TABS.DASHBOARD_TAB],
        lazyLoad: true,
        swipeEnabled: false,
        animationEnabled: false,
        tabBarOptions: {
          showIcon: true,
          style: {
            backgroundColor: Colors.headerBackground,
            height: normalizeValue(65),
            padding: normalizeValue(10)
          },
          labelStyle: textStyles.footerLabel,
          activeTintColor: Colors.brandDefault,
          inactiveTintColor: Colors.defaultText
        }
      }
    );`

A child stack nav:

this.StackNavigator = Router(
      {
        headerComponent: CustomHeader,
        headerMode: 'screen',
        initialRouteName: ROUTES.FIND_TEACHER_TAB,
        navigationBarOptions: {
          header: {
            title: () => i18n.t('home.chatHeaderTitle'),
          }
        }
      }
    );

**Comment by microwavesafe** [comment]
Your right, you can't use set tabBar properties from inside the stack page component. But you can use setParams to set parameters from any route to any route.

```
import { NavigationActions } from 'react-navigation'

const setParamsAction = NavigationActions.setParams({
  params: {}, // these are the new params that will be merged into the existing route params
  // The key of the route that should get the new params
  key: 'screen-123',
})
this.props.navigation.dispatch(setParamsAction)
```

as seen on this docs page

https://reactnavigation.org/docs/navigators/navigation-prop

So you'd have the visible param set up on one of your tab components, inside the tabBar options.
```
visible: !state.params ? true : state.params.visible,
```
then from the stack component call ```this.props.navigation.dispatch(setParamsAction)``` with the correct params and route set. I haven't tested this, so I can't be sure.

**Comment by AlanFoster** [comment]
I'm not quite sure where to post this; But I initially couldn't get the tabBar hiding successfully using this method. In my set up I had configured a top-level TabNavigator with icon configuration, and it wasn't making use of the `navigationOptions` provided within a child within a StackNavigator that had `visible: false`

I was finally able to track down a fix for the issue; I had set `navigationOptions` within the top level TabNavigator, which was ignoring the child's tabBar configuration for some reason. This was my previous TabNavigator configuration:

```js
const DefaultTabNavigation = TabNavigator({
  Main: {
    screen: MainScreen,
    navigationOptions: {
      tabBar: () => ({
        label: 'Home',
        icon: tabIconFor('md-stats')
      })
    }
  }
```

I fixed this by making use of the defaultOptions argument:

```js
const DefaultTabNavigation = TabNavigator({
  Main: {
    screen: MainScreen,
    navigationOptions: {
      // Ensure that the defaultOptions are mixed in too, otherwise the 
      tabBar: (navigation, defaultOptions) => ({
        ...defaultOptions,
        label: 'Home',
        icon: tabIconFor('md-stats')
      })
    }
  }
```

I'm not sure is is a small API oversight or not, i.e. the TabNavigator options should have mixed in with the child's options - rather than replacing the entire configuration from a child view.

**Comment by bhavyasruthi** [comment]
But now "tabBar" is deprecated only tabBarIcon is in use. How to hide tab bar now ?

**Comment by nikasvan** [comment]
Any news about this?

**Pull request merged**